File: algorithm/programmers_정수삼각형.js
function solution(triangle) {
    const values = [...triangle];
    for (let i = 1; i < values.length; i++) {
        for (let j = 0; j < values[i].length; j++) {
            if (j === 0) values[i][j] += values[i-1][j];
            else if (j === values[i].length - 1) values[i][j] += values[i-1][j-1];
            else values[i][j] += Math.max(values[i-1][j-1], values[i-1][j]);
        }
    }
    return Math.max(...values[values.length - 1]);
}

File: algorithm/programmers_정수삼각형.py
# def solution(triangle):
#     for height in range(1, len(triangle)):
#         length = len(triangle[height])
#         for i in range(length):
#             triangle[height][i] += max(triangle[height-1][i-1 if i != 0 else 0], triangle[height-1][i if i != length-1 else -1])

#     return max(triangle[-1])


def solution(triangle):
    for height in range(1, len(triangle)):
        
        triangle[height][0] += triangle[height-1][0]
        for i in range(1, len(triangle[height])-1):
            triangle[height][i] += max(triangle[height-1][i-1], triangle[height-1][i])
        triangle[height][-1] += triangle[height-1][-1]

    return max(triangle[-1])
